src/lib/Autodiff.hs

```diff
@@ -641,6 +636,16 @@ linAtomRef (Var x) = do
   case envLookup refs x of
     Just ref -> return ref
     _ -> error $ "Not a linear var: " ++ pprint (Var x)
+linAtomRef (ProjectElt (i NE.:| is) x) = do
+  let subproj = getProjection is (Var x)
+  case getType subproj of
+    PairTy _ _ -> do
+      ref <- linAtomRef subproj
+      (traverse $ emitOp . getter) ref
+      where getter = case i of 0 -> FstRef
+                               1 -> SndRef
+                               _ -> error "bad pair projection"
+    ty -> error $ "Projecting references not implemented for type " <> pprint ty
```

Comment:
Can't we inline this handling into the `ProjectElt` site below? I'm pretty sure this function was meant to only be used when the `Atom` is a `Var` (at least that's what the error message says below). Do we expect more contangents to be `ProjectElt`s now?

Can we add some tests for this code? I'd also appreciate a comment, because I'm not 100% sure that this implementation is correct. Why do we handle the last level of unpacking specially and treat the rest with `getProjection`?

Reply:
`linAtomRef` is also used when transposing table indexing:
https://github.com/google-research/dex-lang/blob/0acecb3fc567ca8a6e10dc77c86ec9892caa525d/src/lib/Autodiff.hs#L556-L561

This situation happens sometimes when projecting a pair that contains a table, and then indexing into that table.

In practice this is covered by our existing autodiff tests (in the sense that they fail without this case and pass with it). I can try to think of some more targeted tests for this too.

The last-level behavior is essentially doing recursion and handling one level of projection at a time (see the `linAtomRef subproj` subexpression). If we get rid of the `NonEmpty` part, this could probably be simplified. (I'm also being a bit lazy by reusing `getProjection` instead of building the recursive step manually.)

Reply:
Right, ok. I understand why we might want to have it in this function.

Ohhh I see the recursion now. But isn't the `getProjection is (Var x)` part equivalent to saying `ProjectElt is x`? I understand that the difficulty might be that `is` might become empty, in which case you just want to take `Var x`, but honestly I would find that logic much simpler to follow. The way it looks right now seems to imply as if you actually wanted to unpack a value, in a function that's supposed to slice into a reference. This is why I've found it confusing.

Reply:
Yep, that's all it's doing. I'll switch to the explicit form.